src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/WhenShouldBeUsedCheck.java

```diff
@@ -0,0 +1,176 @@
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Ensures that {@code when} is used instead of a single {@code if}
+ * statement inside a case block.
+ * </p>
+ * <p>
+ * Java 21 has introduced enhancements for switch statements and expressions
+ * that allow the use of patterns in case labels. The {@code when} keyword is used to specify
+ * condition for a case label, also called as guarded case labels. This syntax is more readable
+ * and concise than the single {@code if} statement inside the pattern match block.
+ * </p>
+ * <p>
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-Guard">
+ * Java Language Specification</a> for more information about guarded case labels.
+ * </p>
+ * <p>
+ * See the <a href="https://docs.oracle.com/javase/specs/jls/se22/html/jls-14.html#jls-14.30">
+ * Java Language Specification</a> for more information about patterns.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code when.should.be.used}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ */
+
+@FileStatefulCheck
+public class WhenShouldBeUsedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "when.should.be.used";
+
+    /**
+     *  An array of acceptable statements inside the case block.
+     */
+    private static final int[] ACCEPTABLE_STATEMENTS_INSIDE_CASE_BLOCK = {
+        TokenTypes.LITERAL_IF,
+        TokenTypes.LITERAL_BREAK,
+        TokenTypes.EMPTY_STAT,
+        TokenTypes.RCURLY,
+    };
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.LITERAL_CASE};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final DetailAST statementList = getStatementListOfCase(ast);
+        final boolean hasPatternLabel = hasPatternLabel(ast);
+        if (hasPatternLabel) {
+            final List<DetailAST> blockStatements = getBlockStatements(statementList);
```

Comment:
pseudo code:
```suggestion
            final List<DetailAST> blockStatements = statementList.map(WhenShouldBeUsedCheck::getBlockStatements)
                     .orElseGet(List::new);
```

Reply:
I don't think we need this for now. we added `statementList != null` in the if so we don't continue with a null slist. 

Do you think we still should follow your sugesstion?

Reply:
This is good